19274-combox-dropdown

roshaniMali5 opened the pull request:
Closes #19274

{{short description}}

### Changelog


**Changed**

- Fix ai-label.ts: improve popup open/close behavior, focus tracking, and clean up type checks

#### Testing / Reviewing

- Combox with Ai Label

## PR Checklist

<!-- 
  Do not remove checklist items.
  If some are incomplete, create a draft pull request using the create button dropdown.
  If some do not apply, ~strike through the item text with tildes~.
-->

As the author of this PR, before marking ready for review, confirm you:

- [x] Reviewed every line of the diff
- [x] Updated documentation and storybook examples
- [x] Wrote passing tests that cover this change
- [x] Addressed any impact on accessibility (a11y)
- [x] Tested for cross-browser consistency
- [x] Validated that this code is ready for review and status checks should pass

More details can be found in the [pull request guide](https://github.com/carbon-design-system/carbon/blob/main/docs/guides/reviewing-pull-requests.md)

roshaniMali5 committed:
fix: resolve dropdown close issue and added test case

adamalston commented on packages/react/src/components/ComboBox/ComboBox.tsx:
```diff
@@ -894,6 +894,14 @@ const ComboBox = forwardRef(
+      // Close the dropdown when focusing on the AI label or Toggletip
+      if (evt.type === 'focus' && (
+        evt.target.closest(`.${prefix}--ai-label`) ||
+        evt.target.closest(`.${prefix}--toggletip`)
```
Are there alternative methods to determine what's focused besides checking for class selectors? That approach seems fragile.

adamalston commented on packages/react/src/components/ComboBox/ComboBox-test.js:
```diff
@@ -1431,5 +1431,52 @@ describe('ComboBox', () => {
+    document.body.removeChild(toggletipElement);
+  });
+
+// ... existing code ...
```
Is this code comment generative AI output, or what is it referring to?

roshaniMali5 replied:
I was writing a test case and forgot to remove a comment. I'll remove it.
For the unit test case, I'm using Watson AI

adamalston commented on packages/react/src/components/ComboBox/ComboBox-test.js:
```diff
@@ -1431,5 +1431,52 @@ describe('ComboBox', () => {
       expect(findInputNode()).toHaveDisplayValue('Option 2');
     });
+
+
```
This file isn't formatted. Did you see any errors creating the commit? The pre-commit hook should have formatted this file unless commit verfication was bypassed.

adamalston commented on packages/react/src/components/ComboBox/ComboBox-test.js:
```diff
@@ -1431,5 +1431,52 @@ describe('ComboBox', () => {
+    // Create a mock toggletip element and trigger focus event
+    const toggletipElement = document.createElement('div');
+    toggletipElement.className = 'cds--toggletip';
+    document.body.appendChild(toggletipElement);
```
I'm not sure whether you've investigated this approach. Could this testing be done using the actual `ToggleTip` component instead of a mock?

2nikhiltom reviewed (changes requested):
Hey @roshaniMali5  - Thanks for tackling this issue. I think the root cause is that the Enter keypress event bubbles up from the `AI Label's` `ToggletipButton` to the `ComboBox` container, causing both to respond to the same event.

Rather than filtering events in each parent component, could we contain them at the source in AILabel? This would be a single fix that solves it for ComboBox, Dropdown, MultiSelect, and any future parent components.

we could add a new keydown handler in [AILabel component here ](https://github.com/carbon-design-system/carbon/blob/ed28a965aafe7342d5401fd0769bf858740171c7/packages/react/src/components/AILabel/index.tsx#L198)and pass it to [ToggletipButton](https://github.com/carbon-design-system/carbon/blob/ed28a965aafe7342d5401fd0769bf858740171c7/packages/react/src/components/AILabel/index.tsx#L217) as a prop 

```
const handleKeyDown = (event: React.KeyboardEvent) => {
  if (event.key === 'Enter' || event.key === ' ' || event.key === 'Escape') {
    event.stopPropagation(); 
  }
```
};

`<ToggletipButton onKeyDown={handleKeyDown}>`
What do you think of this approach?

roshaniMali5 committed:
feat(combo-box): add support for AI label

roshaniMali5 committed:
Merge branch 'main' into 19274-combox-dropdown

roshaniMali5 committed:
feat: resolved overlapped issue of ai icon

roshaniMali5 committed:
Merge branch 'main' into 19274-combox-dropdown

roshaniMali5 committed:
feat: resolved overlapped issue on ai label

roshaniMali5 committed:
Merge remote-tracking branch 'origin/main' into 19274-combox-dropdown

roshaniMali5 committed:
Merge remote-tracking branch 'origin/main' into 19274-combox-dropdown

roshaniMali5 committed:
Merge remote-tracking branch 'origin/main' into 19274-combox-dropdown

roshaniMali5 committed:
Merge remote-tracking branch 'origin/main' into 19274-combox-dropdown

adamalston commented on packages/web-components/src/components/ai-label/ai-label.ts:
```diff
@@ -70,15 +71,47 @@ class CDSAILabel extends CDSToggleTip {
+      event?.key === ' ' ||
+      event?.key === 'Escape'
+    ) {
+      event?.stopPropagation();
```
Are all of the optional chaining operators necessary?

adamalston commented on packages/web-components/src/components/ai-label/ai-label.ts:
```diff
@@ -70,15 +71,47 @@ class CDSAILabel extends CDSToggleTip {
+  };
+
   protected _handleClick = () => {
+    (document.activeElement as HTMLElement | null)?.blur();
```
You shouldn't need the type assertion. There are type-safe ways to perform this action.

adamalston reviewed (commented):
The checklist in https://github.com/carbon-design-system/carbon/pull/19732#issue-3165671172 is incomplete.

roshaniMali5 committed:
feat: resolve review points

roshaniMali5 committed:
Merge remote-tracking branch 'origin/main' into 19274-combox-dropdown

adamalston reviewed (commented):
> * [x]  Wrote passing tests that cover this change

This task shouldn't be marked as complete. There are no testing changes in this pull request that cover the issue.

roshaniMali5 committed:
Merge remote-tracking branch 'origin/main' into 19274-combox-dropdown

roshaniMali5 committed:
Merge remote-tracking branch 'origin/main' into 19274-combox-dropdown

2nikhiltom commented on packages/web-components/src/components/ai-label/ai-label.ts:
```diff
@@ -70,15 +70,47 @@ class CDSAILabel extends CDSToggleTip {
 
+  @property({ type: Boolean, reflect: true })
+  open = false;
+
```
The open property appears to already exist in parent CDSToggleTip - is the redeclaration intentional for specific behavior, or can we rely on inheritance?
```suggestion

```

2nikhiltom commented on packages/web-components/src/components/ai-label/ai-label.ts:
```diff
@@ -70,15 +70,47 @@ class CDSAILabel extends CDSToggleTip {
+    }
+  };
+
+  protected _handleAIKeydown = (event: React.KeyboardEvent) => {
```
```suggestion
  protected _handleAIKeydown = (event: KeyboardEvent) => {
```

2nikhiltom reviewed (changes requested):
Thanks for putting this PR up, this looks good. just couple of comments

roshaniMali5 committed:
feat: resolve review points

roshaniMali5 committed:
Merge remote-tracking branch 'origin/main' into 19274-combox-dropdown

roshaniMali5 committed:
feat: resolve Tab to focus

roshaniMali5 committed:
Merge remote-tracking branch 'origin/main' into 19274-combox-dropdown

roshaniMali5 committed:
Merge branch 'carbon-design-system:main' into 19274-combox-dropdown

adamalston commented on packages/web-components/src/components/ai-label/ai-label.ts:
```diff
@@ -73,12 +74,56 @@ class CDSAILabel extends CDSToggleTip {
+  };
+
+  private _handleFocusChange = (event: FocusEvent) => {
+    if (this.open && !this.contains(event.target as Node)) {
```
Is there a way to avoid this type assertion?

adamalston reviewed (commented):
> Hello @\adamalston @\2nikhiltom I've resolved this issue. Please review the changes and let me know if everything looks good.

The changes look good to me. I think the description needs to be updated based on https://github.com/carbon-design-system/carbon/pull/19732#issuecomment-3021210615.

2nikhiltom reviewed (approved):
LGTM now

2nikhiltom committed:
Merge branch 'main' into 19274-combox-dropdown

roshaniMali5 committed:
feat: resolve review points

roshaniMali5 committed:
feat: added test case

roshaniMali5 committed:
Merge branch 'carbon-design-system:main' into 19274-combox-dropdown

ariellalgilmore committed:
Merge branch 'main' into 19274-combox-dropdown

ariellalgilmore reviewed (commented):
looks like there are some ts errors causing the build not to finish
<img width="617" height="660" alt="Screenshot 2025-07-21 at 1 11 19 PM" src="https://github.com/user-attachments/assets/e71af443-5b76-4a6e-8d20-e22c4c731ac6" />

roshaniMali5 committed:
fix: unused code

roshaniMali5 committed:
Merge branch 'main' into 19274-combox-dropdown

2nikhiltom committed:
Merge branch 'main' into 19274-combox-dropdown

2nikhiltom reviewed (changes requested):
Hey @roshaniMali5 The test file you recently added differs from the one currently present in the system:
https://github.com/carbon-design-system/carbon/blob/2260989bb0d242d89ec6d5f9a0bc285dd1a397cd/packages/web-components/src/components/ai-label/__tests__/ai-label-test.js#L32

Could you please update the existing one by adding a new test case to verify the fix introduced in this PR

dependabot[bot] committed:
chore(deps): bump form-data in /packages/react/examples/react-router (#19974)

Bumps [form-data](https://github.com/form-data/form-data) from 3.0.1 to 3.0.4.
- [Release notes](https://github.com/form-data/form-data/releases)
- [Changelog](https://github.com/form-data/form-data/blob/v3.0.4/CHANGELOG.md)
- [Commits](https://github.com/form-data/form-data/compare/v3.0.1...v3.0.4)

---
updated-dependencies:
- dependency-name: form-data
  dependency-version: 3.0.4
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Heloise Lui <71858203+heloiselui@users.noreply.github.com>

kennylam committed:
fix(feature-flags): update cjs entrypoint (#19984)

Co-authored-by: Gururaj J <89023023+Gururajj77@users.noreply.github.com>

roshaniMali5 committed:
fix: updated test case of ai-label-test

roshaniMali5 committed:
fix: updated test case

Gururajj77 committed:
fix: changed the color tokens to follow guidelines (#19990)

roshaniMali5 committed:
fix: test case

roshaniMali5 committed:
Merge remote-tracking branch 'origin/main' into 19274-combox-dropdown

roshaniMali5 committed:
fix: resolve test case

roshaniMali5 committed:
Merge branch 'carbon-design-system:main' into 19274-combox-dropdown

roshaniMali5 committed:
Merge branch 'carbon-design-system:main' into 19274-combox-dropdown

roshaniMali5 committed:
Merge branch 'carbon-design-system:main' into 19274-combox-dropdown

ariellalgilmore reviewed (approved):
LGTM
